---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
```

## Setup

```{r}
number_of_genes <- 5
gene_length <- 2000
chromosome <- data.frame(type         = rep(c("gene", "intergenic"), times = number_of_genes), 
                         length       = rep(c(gene_length, 10000),  times = number_of_genes), 
                         essentiality = rep(c(0.95,0.05),            times = number_of_genes), # can we combine the next two into a single measuure? fitness cost of losing
                         intactness   = rep(c(1,1),                  times = number_of_genes)
                         ) 

message("we can eliminate the 'essentiality' vector. instead just use intactness and once it reaches 0, delete the gene")

message("may want to store start/end coordinates instead of length")

#ggplot(chromosome, aes(x=1,y=length, fill=type)) + geom_bar(stat="identity")  # doesn't work as it doesn't maintain the order. 
chromosome
```

# Helper function

* Assumptions unrealistic. 
* "too simplistic - here we choose location randomly by ROW - should be a function of length - may be easier with a "column indicating start positoin)
* Should we use poisson distribution?

```{r}
mutate <- function(current_chromosome, max_insertion_size = 10000, max_deletion_size = 10000) {
   insertion_mutation <- "insertion"
   deletion_mutation  <- "deletion"
   mutation <- sample(x    = c(insertion_mutation, deletion_mutation), 
                      prob = c(0.7, 0.3), 
                      size = 1)
   
   location <- sample(x = 1:nrow(current_chromosome), size = 1)
   message("doing ", mutation, " at ", location, ", a ", current_chromosome$type[location])
   if (mutation == insertion_mutation) {
      size <- sample(x = 1:max_insertion_size, size = 1)  # may want a non-uniform distribution
      current_chromosome$length[location] <- current_chromosome$length[location] + size
      
      if (current_chromosome$type[location] == "gene") {
        current_chromosome$intactness[location] <- 0
      } else {
        # inserting in intergenic. 
        # Change intactness of this one, and regulation (i.e. intactness) of neighbor genes by 5%
        size_deviation <- current_chromosome$length[location]/(chromosome$length[location] + size)
        current_chromosome$intactness[location]  <- current_chromosome$intactness[location] * size_deviation
        if (location != 1) { # edge case for first and last row
          current_chromosome$intactness[location - 1] <- current_chromosome$intactness[location - 1] * 0.95
        }
        if (location != nrow(current_chromosome)) {
          current_chromosome$intactness[location + 1] <- current_chromosome$intactness[location + 1] * 0.95
        }
      }
    } else if(mutation == deletion_mutation) {
      message("   need to implement - supposes two doublestranded breaks -> so can span multiple genes")
    } else {
      die("we should not be here!")
    }
   return(current_chromosome)
}
```


# Do it

```{r}
evolve <- function(current_chromosome, generations = 10) {
  for (i in 1:10) {
    #todo: "if the lineage survived, update essentialities:
    #      for each gene, if intactness was lowered by a small amount, essentiality reduces." (rest of genome adapts)
    
    chromosome <- mutate(current_chromosome = chromosome)
    
       # if any(  intactness == 0 AND essentiality > 0.8)  then lineage dies. 
    
  }
  return(chromosome)
}
evolved_chromosome <- evolve(chromosome)

evolved_chromosome
```

