---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---

# Introduction

```{r functions, echo = FALSE}

library(tidyverse)

source('simulation_functions.R')

results_colnames <- c("simulation_type",
                      "simulation_n",
                      "generation",
                      "mean_length",
                    	"median_length",
                    	"sd_length",
                    	"mad_length",
                    	"max_length",
                    	"min_length",
                    	"mean_value",
                    	"sd_vale",
                    	"median_value",
                    	"mad_value",
                    	"max_value",
                    	"min_value",
                    	"mean_junk",
                    	"sd_junk",
                    	"median_junk",
                    	"mean_mutated_loci",
                    	"median_mutated_loci",
                    	"sd_mutated_loci")
```

Using optical maps and whole-genome sequences, we have shown that the supergene variant with suppressed recombination has increased in size. Our interpretations of these results is that this supergene variant has undergone "degenerative expansion". Our argument is that although non-recombining chromosomes accumulate deleterious mutations, they are expected to accumulate weak deleterious mutations faster than relatively stronger deleterious mutations. Because deletions are thought to be generally more deleterious than insertions, non-recombining chromosomes are expected to increase in size.

We have developed a simulation to illustrate our argument. We simulate a population of 1000 individuals, each carrying a single non-recombining haploid chromosome. Each chromosome is initially comprised of 4000 loci. These repsent protein coding genes and other elements such as regulatory sites. All loci are "functional", being assigned an arbitrary fitness value of 10. There are three types of mutations: deletions, insertions and point mutations. Deletions decrease the size of the chromosome and are deleterious because they remove 'functional' loci. Insertions increase the size of the chromosome. They are not necessarily deleterious, although we add an optional cost to two loci directly neighbouring the insertion (the loci go from having value 10 to value 0). Deletions and insertions have the same fixed size (50 loci). The point mutations represent background degeneration of the chromosome, giving an arbitrary cost to the loci where they affect. For instance, if we choose the point mutation cost to be 2, the value of the locus decreases from 10 to 8. In a given individual, the same locus can be hit by multiple mutations in different generations, thus becoming less fit over time.

Each generation has three steps. In the first step we mutate the chromosome for each individual. The number of insertions and deletions in the chromosome follow a Poisson distribution with the parameter equal to a chosen mutation rate times the size of the chromosome. The number of point mutations is the round of the multiplication of the point mutation rate by the size of the chromosome. The locations of the mutations are sampled randomly from the chromosome without replacement. For deletions, we remove the affected locu. For insertions, we reduce the value of the two loci neighbouring the insertion location by the insertion cost, and add loci valued 0 between them. For the point mutations, we reduce the value of the affected loci by the point mutation cost.

In the second step, we assign a fitness to each individual as the sum of all the loci values. We determin the relative fitness of each individual by dividing their fitness by the fitness of the individual with highest fitness.

In the third step, we select 1000 individuals to form the next generation by sampling them, with replacement, with probability equal to their relative fitnesses.


## Run simulations

We first test cases with neutral evolution. We care at looking at different rates of insertions and deletions relative to each other.

```{r neutral_simulation, cache=TRUE, include=TRUE, echo = FALSE}

neutral_simulations <- read.table("tmp/simulations/neutral_all_sim")

colnames(neutral_simulations) <- results_colnames

neutral_simulations$simulation_n <- as.character(neutral_simulations$simulation_n)
neutral_simulations$simulation_n <- gsub("\\.txt", "", neutral_simulations$simulation_n)
neutral_simulations$simulation_n <- gsub("simulation_", "", neutral_simulations$simulation_n)
neutral_simulations$simulation_n <- as.numeric(neutral_simulations$simulation_n)

neutral_simulations$simulation         <- neutral_simulations$simulation_type
levels(neutral_simulations$simulation) <- c("Neutral, no bias",
                                            "Neutral, higher deletion rate",
                                            "Neutral, higher insertion rate")

neutral_simulations %>%
  filter(generation <= 1000 & simulation_n %in% 1:5) %>%
  ggplot() +
  geom_line(aes(x = generation,
                y = mean_length,
                colour = simulation,
                group = interaction(simulation, simulation_n)),
            size = 0.5) + 
  theme_bw() + theme(legend.position="right")

```

We now test whether accruing a cost to deletions increases the size of the chromosome. To do this, we give each locus a fitness value (fixed at 10), which is removed by deletions. Insertions have either no fitness cost, or a fitness cost affecting only the two loci neighbouring each insertion (which is a lower fitness cost than deletions).

```{r deletion_cost, include=TRUE, echo = FALSE, cache=FALSE}

deletion_cost <- read.table("tmp/simulations/deletion_with_cost_all_sim")

colnames(deletion_cost) <- results_colnames

deletion_cost$simulation_n <- as.character(deletion_cost$simulation_n)
deletion_cost$simulation_n <- gsub("\\.txt", "", deletion_cost$simulation_n)
deletion_cost$simulation_n <- gsub("simulation_", "", deletion_cost$simulation_n)
deletion_cost$simulation_n <- as.numeric(deletion_cost$simulation_n)

deletion_cost$simulation <- deletion_cost$simulation_type
levels(deletion_cost$simulation) <- c("Deleterious deletions,\n  deleterious insertions",
                                      "Deleterious deletions,\n  neutral insertions" )


deletion_cost            <- rbind(deletion_cost,
                                  subset(neutral_simulations, simulation == "Neutral, no bias"))

re_level <- levels(deletion_cost$simulation) == "Neutral, no bias"
deletion_cost$simulation <- factor(deletion_cost$simulation,
                                     levels = levels(deletion_cost$simulation)[c(3,2,1)])

deletion_cost %>%
  filter(generation <= 1000 & simulation_n %in% 1:5) %>%
  ggplot() +
  geom_line(aes(x = generation,
                y = mean_length,
                colour = simulation,
                group = interaction(simulation, simulation_n)),
            size = 0.5) +
  theme_bw() + theme(legend.position="right")

```
The results of the simulations start to get a bit noisy, but we can solve this using a smoothing function.

```{r deletion_cost_smooth, include=TRUE, echo = FALSE, cache=FALSE}

deletion_cost %>%
    filter(generation <= 1000 & simulation_n %in% 1:5) %>%
  ggplot() +
  stat_smooth(aes(x = generation,
                  y = mean_length,
                  colour = simulation),
              size = 0.5) + 
  theme_bw() + theme(legend.position="right")

```

Our simulations suggest that the size of the chromosome undergoes a very large increase, but slower to decrease. This is due to the lower of a cost of insertions relative to deletions. The size increase seems to plateau, as deletions become more likely to affect previously inserted sites, where they are no longer deleterious.

## Effect of background degeneration

However, a non-recombining chromosome is thought to undergo sequence-level degeneration as well degeneration based on insertions and deletions. In the following simulation, we include weakly deleterious point mutations that decrease the fitness value of the locus (which starts at 10) by 2. In this simulation, the size of the chromosome eventually plateaus, as the cost of deletions decreases. We do not include a cost for insertions in these simulations.

```{r deletion_point_mutation, include=TRUE, echo = FALSE, cache=FALSE}

point_mutations <- read.table("tmp/simulations/point_mutations_all_sim")

colnames(point_mutations) <- results_colnames

point_mutations$simulation_n <- as.character(point_mutations$simulation_n)
point_mutations$simulation_n <- gsub("\\.txt", "", point_mutations$simulation_n)
point_mutations$simulation_n <- gsub("simulation_", "", point_mutations$simulation_n)
point_mutations$simulation_n <- as.numeric(point_mutations$simulation_n)

point_mutations$simulation <- point_mutations$simulation_type
levels(point_mutations$simulation) <- c("Deleterious deletions,\n background degeneration",
                                        "Deleterious deletions,\n stronger background degeneration")

point_mutations <- rbind(point_mutations, subset(deletion_cost, simulation == "Deleterious deletions,\n  neutral insertions"))
point_mutations$simulation <- droplevels(point_mutations$simulation)

re_level <- levels(point_mutations$simulation) == "Deleterious deletions,\n  neutral insertions"
levels(point_mutations$simulation)[re_level] <- "Deleterious deletions,\n no background degeneration"
point_mutations$simulation <- factor(point_mutations$simulation,
                                     levels = levels(point_mutations$simulation)[c(which(re_level), which(!re_level))])

point_mutations %>%
  filter(generation <= 1000 & simulation_n %in% 1:5) %>%
  ggplot() +
    geom_line(aes(x = generation,
                  y = mean_length,
                  colour = simulation,
                  group = interaction(simulation, simulation_n)),
              size = 0.5) + 
    theme_bw() + theme(legend.position="right")

```


The results of the simulations start to get a bit noisy, but we can solve this using a smoothing function.

```{r deletion_point_mutation_smooth, include=TRUE, echo = FALSE, cache=FALSE}

point_mutations %>%
    filter(generation <= 1000 & simulation_n %in% 1:5) %>%
  ggplot() +
  stat_smooth(aes(x = generation,
                  y = mean_length,
                  colour = simulation),
              size = 0.5) + 
  theme_bw() + theme(legend.position="right")

```

Adding background degeneration slows down the initial increase, but without causing the chromosome to decrease again.

## Can the chromosome size decrease again?

There are two conditions in which the size of the chromosomes can decrease after their initial increase. The first a having a deletion rate that is slightly higher than the insertion rate, which slows down the initial increase in size of the non-recombining chromosome, but eventually accelerates its decrease. The other is having cost associated with 'junk' loci in the genome. We simulated this by adding a multiplier cost that reduces an individual's fitness proportionally to the number of non-functional loci (i.e. the loci introduced by insertions or rendered non-functional by repeated point mutations).


```{r insertion_cost, include=TRUE, echo = FALSE, cache=FALSE}

insertion_cost <- read.table("tmp/simulations/insertion_cost_all_sim")

colnames(insertion_cost) <- results_colnames

insertion_cost$simulation_n <- as.character(insertion_cost$simulation_n)
insertion_cost$simulation_n <- gsub("\\.txt", "", insertion_cost$simulation_n)
insertion_cost$simulation_n <- gsub("simulation_", "", insertion_cost$simulation_n)
insertion_cost$simulation_n <- as.numeric(insertion_cost$simulation_n)

insertion_cost$simulation <- insertion_cost$simulation_type
levels(insertion_cost$simulation) <- c("Higher deletion rate",
                                       "Higher deletion rate\no background degeneration",
                                       "Low cost for junk",
                                       "High cost for junk")

insertion_cost <- rbind(insertion_cost,
                        subset(point_mutations, simulation == "Deleterious deletions,\n background degeneration"))

insertion_cost$simulation <- droplevels(insertion_cost$simulation)

re_level <- levels(insertion_cost$simulation) == "Deleterious deletions,\n background degeneration"
insertion_cost$simulation <- factor(insertion_cost$simulation,
                                     levels = levels(insertion_cost$simulation)[c(which(re_level),
                                                                                  which(!re_level))])

insertion_cost %>%
  filter(generation <= 1000 & simulation_n %in% 1:5) %>%
  ggplot() +
  geom_line(aes(x = generation,
                y = mean_length,
                colour = simulation,
                group = interaction(simulation, simulation_n)),
            size = 0.5) + 
  theme_bw() + theme(legend.position="right")

insertion_cost %>%
  filter(generation <= 1000 & simulation_n %in% 1:5) %>%
  ggplot() +
  stat_smooth(aes(x = generation,
                  y = mean_length,
                  colour = simulation),
              size = 0.5) + 
  theme_bw() + theme(legend.position="right")

```
